graphbolt/src/concurrent_id_hash_map.cc

```diff
@@ -199,6 +202,29 @@ inline void ConcurrentIdHashMap<IdType>::InsertAndSet(IdType id, IdType value) {
   hash_map_.data_ptr<IdType>()[getValueIndex(pos)] = value;
 }
 
+template <typename IdType>
+void ConcurrentIdHashMap<IdType>::InsertAndSetMin(IdType id, IdType value) {
+  IdType pos = (id & mask_), delta = 1;
+  IdType* hash_map_data = hash_map_.data_ptr<IdType>();
+  InsertState state = AttemptInsertAt(pos, id);
+  while (state == InsertState::OCCUPIED) {
+    Next(&pos, &delta);
+    state = AttemptInsertAt(pos, id);
+  }
+
+  IdType empty_key = static_cast<IdType>(kEmptyKey);
+  IdType val_pos = getValueIndex(pos);
+  IdType old_val = empty_key;
+  while (old_val == empty_key || old_val > value) {
+    IdType replaced_val =
+        CompareAndSwap(&(hash_map_data[val_pos]), old_val, value);
+    if (old_val == replaced_val)
+      break;
+    else
+      old_val = replaced_val;
```

Comment:
I would take the changes here back, there is no reason to make it more complicated. The performance won't change either.
```suggestion
    old_val = CompareAndSwap(&(hash_map_data[val_pos]), old_val, value);
```

Reply:
But in the old code, after a successful `CompareAndSwap`, the loop won't stop, it will do an extra `CompareAndSwap`, which could be slower

Reply:
Makes sense, good catch! How much faster is this new implementation? We can do this because CompareAndSwap is the https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange strong variant, right? It is guaranteed to succeed if there is no contention.